README.md
<p align="center">
  <img src="https://getkap.co/static/favicon/kap.svg" height="64">
  <h3 align="center">KapMod</h3>
  <p align="center">An open-source screen recorder built with web technology</p>
  <p align="center"><a href="https://circleci.com/gh/wulkano/kap">
    <a href="https://github.com/sindresorhus/xo">
      <img src="https://img.shields.io/badge/code_style-XO-5ed9c7.svg" alt="XO code style">
    </a>
  </p>
</p>

### Is this mod working ?

Yep !! But his is a modification on top of Kap. Make sure you like it first, then you can make your own functions :)

  OS: big sur
  
  Fork from Kap3.3.2 - 10th Sep 2020
  
  [Aperture is a swift lib for MacOs](https://github.com/wulkano/Aperture)
  
Upstream Kap's issue:

[Can not pause](https://github.com/wulkano/aperture-node/issues/5)

### Instalation Steps:

  - Read `How to build` section

  - Especially, Copy Kap.app to Application

  - Fix issues in [issues](https://github.com/ilovejs/KapMod/blob/main/README.md#issue-fix-log) section

### Motivation - WHY ?

- Low CPU Fan RPM ! 

  Normally, recording in Chrome + Movavi (for example) bump 2/4 CPU to 80%... Fan burst to max level.
  
  With Kap + Safari, you can spend less energy and CPU cycle to record video. say 15mins, the fan slightly pick up at low-middle rpm.
  
  Try to show Fan RPM and CPU temperature:
  ```
  $ sudo powermetrics --samplers smc
  $ sudo powermetrics --samplers smc |grep -i "CPU die temperature"
  ```

- Kap is open source screen recording app on Mac. It's elegant. I want to make use of its UI (Electron)

- I made a default path so no Export waiting time. No convertion when export.

- Aperture is a swift lib for MacOs. Kap use it as binary and js-binding.
  We can optimized parameter as practice.
  
  2k - 4k rpm, 50 degree (+20 room temperature)
  
### WHY NOT ?

[Upper souce library no API for Pause Recording](https://github.com/wulkano/aperture-node/issues/5)

Require swift skills & time..


## Get original Kap to learn GUI

**[Download the latest release](https://kap.now.sh/api)** (macOS only)

Or install with [Homebrew-Cask](https://caskroom.github.io):

```
$ brew install --cask kap
```

## Plugins

For more info on how to create plugins, read the [plugins docs](docs/plugins.md).

## Dev builds

Download [`master`](https://kap-artifacts.now.sh/master) or builds for any other branch using: `https://kap-artifacts.now.sh/<branch>`. Note that these builds are unsupported and may have issues.<br>

## Related Repositories

- [Website](https://github.com/wulkano/kap-website)
- [Aperture](https://github.com/wulkano/aperture)

## Newsletter

[Subscribe](http://eepurl.com/ch90_1)

## Thanks

- [▲ Vercel](https://vercel.com/) for fast deployments served from the edge, hosting our website, downloads, and updates.
- [● CircleCI](https://circleci.com/) for supporting the open source community and making our builds fast and reliable.
- [△ Sentry](https://sentry.io/) for letting us know when Kap isn't behaving and helping us eradicate said behaviour.
- Our [contributors](https://github.com/wulkano/kap/contributors) who help maintain Kap and make screen recording and sharing easy.

## New binding for aperture

Looking at the aperture inside `main`, I put a executable file and index.js for module config.

Files:
  index.js
  main/with reference by aperture.js
  device.js

```
yarn remove aperture
yarn add -D file-url
```

Making sure Intelij IDE is used for developing. Since vscode would quit after run, so as from CLI . Bash.

### How to build

`bash build.sh`

Copy Kap.app to Application !!

### TODO:

tempy is still used by recording-history and convert

we can further extend !!

[1] CICD

https://circleci.com/

appcenter

### Issue fix log

[1] spawn error ENMOT

```md
Recording error after setCurrentRecording()
Error: Command failed with ENOENT: /Applications/Kap.app/Contents/Resources/app.asar.unpacked/main/utils/aperture/aperture {"destination":"file:///Users/mike/Movies/Kaptures/a3d6294fb5fae827b4a5a45723412b18.mp4","framesPerSecond":30,"showCursor":false,"highlightClicks":false,"screenId":69733632,"audioDeviceId":"AppleHDAEngineInput:1F,3,0,1,0:1","cropRect":[[419,379],[227,209]]}
spawn /Applications/Kap.app/Contents/Resources/app.asar.unpacked/main/utils/aperture/aperture ENOENT
```

Again, this is `spawn` issue,

* Quick Fix:

U can create subfolder `main/utils/aperture/` and copy binary `aperture` there.

Binary is available after install from yarn, visit node_modules/aperture ...

* Proper Solution: Need to publish `aperture` as separated npm package

like others:
Kap.app/Contents/Resource/app.asar.unpacked/node_modules:
  file-icon




changes.txt


## Get Kap

**[Download the latest release](https://kap.now.sh/api)** (macOS only)

Or install with [Homebrew-Cask](https://caskroom.github.io):

```
$ brew install --cask kap
```

## Contribute

Read the [contribution guide](contributing.md).

## Plugins

For more info on how to create plugins, read the [plugins docs](docs/plugins.md).

## Dev builds

Download [`master`](https://kap-artifacts.now.sh/master) or builds for any other branch using: `https://kap-artifacts.now.sh/<branch>`. Note that these builds are unsupported and may have issues.<br>

## Related Repositories

- [Website](https://github.com/wulkano/kap-website)
- [Aperture](https://github.com/wulkano/aperture)

## Newsletter

[Subscribe](http://eepurl.com/ch90_1)

## Thanks

- [▲ Vercel](https://vercel.com/) for fast deployments served from the edge, hosting our website, downloads, and updates.
- [● CircleCI](https://circleci.com/) for supporting the open source community and making our builds fast and reliable.
- [△ Sentry](https://sentry.io/) for letting us know when Kap isn't behaving and helping us eradicate said behaviour.
- Our [contributors](https://github.com/wulkano/kap/contributors) who help maintain Kap and make screen recording and sharing easy.




main/utils/devices.js
'use strict';

const audioDevices = require('macos-audio-devices');

// const aperture = require('aperture');
const aperture = require('./aperture')

const {showError} = require('./errors');
const {hasMicrophoneAccess} = require('../common/system-permissions');

const getAudioDevices = async () => {
  if (!hasMicrophoneAccess()) {
    return [];
  }

  try {
    const devices = await audioDevices.getInputDevices();

    return devices.sort((a, b) => {
      if (a.transportType === b.transportType) {
        return a.name.localeCompare(b.name);
      }

      if (a.transportType === 'builtin') {
        return -1;
      }

      if (b.transportType === 'builtin') {
        return 1;
      }

      return 0;
    }).map(device => ({id: device.uid, name: device.name}));
  } catch (error) {
    try {
      const devices = await aperture.audioDevices();

      if (!Array.isArray(devices)) {
        const Sentry = require('./sentry');
        Sentry.captureException(new Error(`devices is not an array: ${JSON.stringify(devices)}`));
        showError(error);
        return [];
      }

      return devices;
    } catch (error) {
      showError(error);
      return [];
    }
  }
};

const getDefaultInputDevice = () => {
  try {
    const device = audioDevices.getDefaultInputDevice.sync();
    return {
      id: device.uid,
      name: device.name
    };
  } catch {
    // Running on 10.13 and don't have swift support libs. No need to report
  }
};

module.exports = {getAudioDevices, getDefaultInputDevice};


main/common/aperture.js
'use strict';

// require('aperture')
const createAperture = require("../utils/aperture");

const {openEditorWindow} = require('../editor');
const {closePrefsWindow} = require('../preferences');
const {setRecordingTray, disableTray, resetTray} = require('../tray');
const {disableCroppers, setRecordingCroppers, closeAllCroppers} = require('../cropper');
const {setCropperShortcutAction} = require('../global-accelerators');

// eslint-disable-next-line no-unused-vars
const {convertToH264} = require('../utils/encoding');

const settings = require('./settings');
const {track} = require('./analytics');
const plugins = require('./plugins');
const {getAudioDevices} = require('../utils/devices');
const {showError} = require('../utils/errors');
const {RecordServiceContext} = require('../service-context');
const {setCurrentRecording, updatePluginState, stopCurrentRecording} = require('../recording-history');

const aperture = createAperture();
const {videoCodecs} = createAperture;

// eslint-disable-next-line no-unused-vars
const recordHevc = videoCodecs.has('hevc');

let lastUsedSettings;
let recordingPlugins = [];
const serviceState = new Map();
let apertureOptions;
let recordingName;
let past;

const setRecordingName = name => {
  recordingName = name;
};

const serializeEditPluginState = () => {
  const result = {};

  for (const {plugin, service} of recordingPlugins) {
    if (!result[plugin.name]) {
      result[plugin.name] = {};
    }

    result[plugin.name][service.title] = serviceState.get(service.title).persistedState;
  }

  return result;
};

const callPlugins = async method => Promise.all(recordingPlugins.map(async ({plugin, service}) => {
  if (service[method] && typeof service[method] === 'function') {
    try {
      await service[method](
        new RecordServiceContext({
          apertureOptions,
          state: serviceState.get(service.title),
          config: plugin.config,
          setRecordingName
        })
      );
    } catch (error) {
      showError(error, {title: `Something went wrong while using the plugin “${plugin.prettyName}”`, plugin});
    }
  }
}));

const cleanup = async () => {
  closeAllCroppers();
  resetTray();

  await callPlugins('didStopRecording');
  serviceState.clear();

  setCropperShortcutAction();
};

const startRecording = async options => {
  if (past) {
    return;
  }

  past = Date.now();
  recordingName = undefined;

  closePrefsWindow();
  disableTray();
  disableCroppers();

  const {cropperBounds, screenBounds, displayId} = options;

  cropperBounds.y = screenBounds.height - (cropperBounds.y + cropperBounds.height);

  const {
    record60fps,
    showCursor,
    highlightClicks,
    recordAudio
  } = settings.store;

  apertureOptions = {
    fps: record60fps ? 60 : 30,
    cropArea: cropperBounds,
    showCursor,
    highlightClicks,
    screenId: displayId
  };

  lastUsedSettings = {
    recordedFps: apertureOptions.fps
  };

  if (recordAudio === true) {
    // In case for some reason the default audio device is not set
    // use the first available device for recording
    const audioInputDeviceId = settings.getSelectedInputDeviceId();
    if (audioInputDeviceId) {
      apertureOptions.audioDeviceId = audioInputDeviceId;
    } else {
      const [defaultAudioDevice] = await getAudioDevices();
      apertureOptions.audioDeviceId = defaultAudioDevice && defaultAudioDevice.id;
    }
  }

  // TODO: figure out how to correctly process hevc videos with ffmpeg
  // if (recordHevc) {
  //   apertureOptions.videoCodec = 'hevc';
  // }

  console.log(`Collected settings after ${(Date.now() - past) / 1000}s`);

  recordingPlugins = plugins
    .getRecordingPlugins()
    .flatMap(
      plugin => plugin.recordServicesWithStatus
        // Make sure service is valid and enabled
        .filter(({title, isEnabled}) => isEnabled && plugin.config.validServices.includes(title))
        .map(service => ({plugin, service}))
    );

  for (const {service, plugin} of recordingPlugins) {
    serviceState.set(service.title, {persistedState: {}});
    track(`plugins/used/record/${plugin.name}`);
  }

  await callPlugins('willStartRecording');

  console.log("callPlugins");

  try {
    const filePath = await aperture.startRecording(apertureOptions);
    console.log("startRecording() is done")
    console.log("filePath: ", filePath);
    console.log("Run setCurrentRecording()")

    setCurrentRecording({
      filePath,
      name: recordingName,
      apertureOptions,
      editPlugins: serializeEditPluginState()
    });

    console.log("setCurrentRecording is ok");

  } catch (error) {
    track('recording/stopped/error');

    showError(error, {title: 'Recording error after setCurrentRecording()'});

    console.log("error", error)

    past = null;
    cleanup();
    return;
  }

  const startTime = (Date.now() - past) / 1000;
  if (startTime > 3) {
    track(`recording/started/${startTime}`);
  } else {
    track('recording/started');
  }

  console.log(`Started recording after ${startTime}s`);
  setRecordingCroppers();
  setRecordingTray(stopRecording);
  setCropperShortcutAction(stopRecording);
  past = Date.now();

  // Track aperture errors after recording has started, to avoid kap freezing if something goes wrong
  aperture.recorder.catch(error => {
    // Make sure it doesn't catch the error of ending the recording
    if (past) {
      track('recording/stopped/error');
      showError(error, {title: 'Recording error'});
      past = null;
      cleanup();
    }
  });

  await callPlugins('didStartRecording');
  updatePluginState(serializeEditPluginState());
};

const stopRecording = async () => {
  // Ensure we only stop recording once
  if (!past) {
    return;
  }

  console.log(`Stopped recording after ${(Date.now() - past) / 1000}s`);
  past = null;

  let filePath;

  try {
    filePath = await aperture.stopRecording();
  } catch (error) {
    track('recording/stopped/error');
    showError(error, {title: 'Recording error'});
    cleanup();
    return;
  }

  const {recordedFps} = lastUsedSettings;

  try {
    cleanup();
  } finally {
    track('editor/opened/recording');

    // TODO: bring this back when we figure out how to convert hevc files
    // if (recordHevc) {
    //   openEditorWindow(await convertToH264(filePath), {recordedFps, isNewRecording: true, originalFilePath: filePath});
    // } else {
    await openEditorWindow(filePath, {recordedFps, isNewRecording: true, recordingName});
    // }

    stopCurrentRecording(recordingName);
  }
};

module.exports = {
  startRecording,
  stopRecording,
  getAudioDevices
};


main/utils/aperture/index.js
'use strict';
const os = require('os');
const util = require('util');
const path = require('path');
const execa = require('execa');
const macosVersion = require('macos-version');
const fileUrl = require('file-url');
const electronUtil = require('electron-util/node');

const uniqueString = require('unique-string');

const debuglog = util.debuglog('aperture');

// Workaround for https://github.com/electron/electron/issues/9459
const BIN = path.join(electronUtil.fixPathForAsarUnpack(__dirname), 'aperture');

const supportsHevcHardwareEncoding = (() => {
  if (!macosVersion.isGreaterThanOrEqualTo('10.13')) {
    return false;
  }

  // Get the Intel Core generation, the `4` in `Intel(R) Core(TM) i7-4850HQ CPU @ 2.30GHz`
  // More info: https://www.intel.com/content/www/us/en/processors/processor-numbers.html
  const result = /Intel.*Core.*i(?:7|5)-(\d)/.exec(os.cpus()[0].model);

  // Intel Core generation 6 or higher supports HEVC hardware encoding
  return result && Number(result[1]) >= 6;
})();

class Aperture {
  constructor() {
    macosVersion.assertGreaterThanOrEqualTo('10.12');
  }

  startRecording({
                   fps = 30,
                   cropArea = undefined,
                   showCursor = true,
                   highlightClicks = false,
                   screenId = 0,
                   audioDeviceId = undefined,
                   videoCodec = undefined
                 } = {}) {
    return new Promise((resolve, reject) => {
      if (this.recorder !== undefined) {
        reject(new Error('Call `.stopRecording()` first'));
        return;
      }

      // this.tmpPath = tempy.file({extension: 'mp4'});
      const homedir = os.homedir();
      this.tmpPath = homedir + "/Movies/Kaptures/" + uniqueString() + ".mp4"

      if (highlightClicks === true) {
        showCursor = true;
      }

      if (typeof cropArea === 'object') {
        if (typeof cropArea.x !== 'number' ||
          typeof cropArea.y !== 'number' ||
          typeof cropArea.width !== 'number' ||
          typeof cropArea.height !== 'number') {
          reject(new Error('Invalid `cropArea` option object'));
          return;
        }
      }

      const recorderOpts = {
        destination: fileUrl(this.tmpPath),
        framesPerSecond: fps,
        showCursor,
        highlightClicks,
        screenId,
        audioDeviceId
      };

      if (cropArea) {
        recorderOpts.cropRect = [
          [cropArea.x, cropArea.y],
          [cropArea.width, cropArea.height]
        ];
      }

      if (videoCodec) {
        const codecMap = new Map([
          ['h264', 'avc1'],
          ['hevc', 'hvc1'],
          ['proRes422', 'apcn'],
          ['proRes4444', 'ap4h']
        ]);

        if (!supportsHevcHardwareEncoding) {
          codecMap.delete('hevc');
        }

        if (!codecMap.has(videoCodec)) {
          throw new Error(`Unsupported video codec specified: ${videoCodec}`);
        }

        recorderOpts.videoCodec = codecMap.get(videoCodec);
      }

      this.recorder = execa(BIN, [JSON.stringify(recorderOpts)]);

      const timeout = setTimeout(() => {
        // `.stopRecording()` was called already
        if (this.recorder === undefined) {
          return;
        }

        const err = new Error('Could not start recording within 5 seconds');
        err.code = 'RECORDER_TIMEOUT';
        this.recorder.kill();
        delete this.recorder;
        reject(err);
      }, 10000);

      this.recorder.catch(error => {
        clearTimeout(timeout);
        delete this.recorder;
        reject(error);
      });

      this.recorder.stdout.setEncoding('utf8');

      this.recorder.stdout.on('data', data => {

        console.log('data:', data)

        debuglog(data);

        if (data.trim() === 'R') {
          // `R` is printed by Swift when the recording **actually** starts
          clearTimeout(timeout);
          resolve(this.tmpPath);
        }
      });
    });
  }

  async stopRecording() {
    if (this.recorder === undefined) {
      throw new Error('Call `.startRecording()` first');
    }

    console.log("stopRecording...")

    this.recorder.kill();
    await this.recorder;
    delete this.recorder;

    return this.tmpPath;
  }
}

module.exports = () => new Aperture();

module.exports.screens = async () => {
  const stderr = await execa.stderr(BIN, ['list-screens']);

  try {
    return JSON.parse(stderr);
  } catch (_) {
    return stderr;
  }
};

module.exports.audioDevices = async () => {
  const stderr = await execa.stderr(BIN, ['list-audio-devices']);

  try {
    return JSON.parse(stderr);
  } catch (_) {
    return stderr;
  }
};

Object.defineProperty(module.exports, 'videoCodecs', {
  get() {
    const codecs = new Map([
      ['h264', 'H264'],
      ['hevc', 'HEVC'],
      ['proRes422', 'Apple ProRes 422'],
      ['proRes4444', 'Apple ProRes 4444']
    ]);

    if (!supportsHevcHardwareEncoding) {
      codecs.delete('hevc');
    }

    return codecs;
  }
});
